### ISSUE TITLE
"Failure when receiving data from the peer" error from an API client after upgrading from 5.0.2 to 5.3.1

### ISSUE THREAD

**Issue by ndbroadbent** [puma/puma#2632]
I updated puma from `5.0.2` to `5.3.1`, and I'm seeing an unusual crash in one of my e2e integration tests. This test stands up the Rails application, and then sends some requests from a Ruby API client library. It consistently fails in the same place:

```
     Failure/Error:
       fail ApiError.new(:code => 0,
                         :message => response.return_message)
     MyApp::ApiError:
       Failure when receiving data from the peer
```

I also run e2e tests for API clients in other languages (C#, Java, JavaScript, Python, PHP), and all of these tests are fine. The other Ruby client tests are also fine, so it's just this one failure for the Ruby API client. This also doesn't happen on my local machine (a Mac), it only happens in the CI build (Debian.)

Could this be related to Keepalive connections exceeding a limit? (#2627.) It's a very specific single failure, but I just want to make sure that this won't start causing any problems in production. Are there any recent changes that might be causing this behavior? Is this an option that I could configure, and if so, would you recommend configuring it only for the test environment? Thanks!

**Comment by dentarg** [comment]
Would be easier to say something about this if it could be reproduced. Even if you can’t reproduce it on your Mac, can share code that looks the same as your CI and fails in Debian?

**Comment by ndbroadbent** [comment]
Sorry about not providing a PoC, I’m still not sure why this was happening or how to reproduce. But I can confirm that this seems to be fixed on the master branch, so the PR that fixed #2627 probably fixed this issue with my test suite. Thanks!
